lib/galaxy/objectstore/irods.py

```diff
@@ -0,0 +1,676 @@
+
+        try:
+            data_obj = self.session.data_objects.get(data_object_path)
+        except (DataObjectDoesNotExist, CollectionDoesNotExist):
+            log.warn("Collection or data object (%s) does not exist", data_object_path)
+            return False
+
+        if self.cache_size > 0 and data_obj.__sizeof__() > self.cache_size:
+            log.critical("File %s is larger (%s) than the cache size (%s). Cannot download.",
+                         rel_path, data_obj.__sizeof__(), self.cache_size)
+            return False
+
+        log.debug("Pulled data object '%s' into cache to %s", rel_path, self._get_cache_path(rel_path))
+
+        with data_obj.open('r') as data_obj_fp, open(self._get_cache_path(rel_path), "wb") as cache_fp:
+            content = data_obj_fp.read()
+            cache_fp.write(content)
+        return True
+
+    def _push_to_irods(self, rel_path, source_file=None, from_string=None):
+        """
+        Push the file pointed to by ``rel_path`` to the iRODS. Extract folder name
+        from rel_path as iRODS collection name, and extract file name from rel_path
+        as iRODS data object name.
+        If ``source_file`` is provided, push that file instead while
+        still using ``rel_path`` for collection and object store names.
+        If ``from_string`` is provided, set contents of the file to the value of the string.
+        """
+        p = Path(rel_path)
+        data_object_name = p.stem + p.suffix
+        subcollection_name = p.parent
+
+        source_file = source_file if source_file else self._get_cache_path(rel_path)
+        options = {kw.FORCE_FLAG_KW: ''}
+
+        if os.path.exists(source_file):
+            # Check if the data object exists in iRODS
+            collection_path = self.home + "/" + str(subcollection_name)
+            data_object_path = collection_path + "/" + str(data_object_name)
+            exists = self.session.data_objects.exists(data_object_path)
+            if os.path.getsize(source_file) == 0 and exists:
+                log.debug("Wanted to push file '%s' to iRODS collection '%s' but its size is 0; skipping.", source_file, rel_path)
+                return True
+            if from_string:
+                data_obj = self.session.data_objects.create(data_object_path, self.resource, **options)
+                with data_obj.open('w') as data_obj_fp:
+                    data_obj_fp.write(from_string)
+                log.debug("Pushed data from string '%s' to collection '%s'", from_string, data_object_path)
+            else:
+                start_time = datetime.now()
+                log.debug("Pushing cache file '%s' of size %s bytes to collection '%s'", source_file, os.path.getsize(source_file), rel_path)
+
+                # Create sub-collection first
+                self.session.collections.create(collection_path, recurse=True)
+
+                file_content = None
+                with open(source_file, 'rb') as content_file:
```

Comment:
This shouldn't all be in memory right? This will routinely read multi-gigabyte files in production I suspect. Can we set a buffer size and merge the loops into one?

Reply:
That applies to https://github.com/galaxyproject/galaxy/pull/9438/files#diff-b64aafa89fa13b4e704cc68ff34b2da1R373 as well.

Reply:
@jmchilton Sure. Will do. Do you have an example buffered read in Galaxy that I can template this on?

Reply:
@mvdbeek your link brings me back to this PR! Could you please clarify what you are referring to? Thx

Reply:
Sorry, line 374 which should also use chunks.
To read/write in chunks do
```python
from functools import partial
...
CHUNK_SIZE = 2**20 # 1MB

...
with open(source_file, 'rb') as content_file, data_obj.open('w') as data_obj_fp:
    for chunk in iter(partial(content_file.read, CHUNK_SIZE), b''):
        data_obj_fp.write(chunk)
```

You can limit the number of bytes to read with an integer argument to `read`. You know you've reached the end of the file when read() returns nothing, so a more old-school implementation would be:
```python
chunk = content_file.read(CHUNK_SIZE)
while chunk:
    data_obj_fp.write(chunk)
    chunk = content_file.read(CHUNK_SIZE)